packages/file_selector/file_selector_android/android/src/main/java/dev/flutter/packages/file_selector_android/FileUtils.java

```diff
@@ -206,4 +216,29 @@ private static String getBaseName(String fileName) {
     // Basename is everything before the last '.'.
     return fileName.substring(0, lastDotIndex);
   }
+
+  // From https://developer.android.com/privacy-and-security/risks/untrustworthy-contentprovider-provided-filename#sanitize-provided-filenames.
+  protected static @Nullable String sanitizeFilename(@Nullable String displayName) {
+    if (displayName == null) {
+      return null;
+    }
+
+    String[] badCharacters = new String[] { "..", "/" };
+    String[] segments = displayName.split("/");
+    String fileName = segments[segments.length - 1];
+    for (String suspString : badCharacters) {
+      fileName = fileName.replace(suspString, "_");
+    }
+    return fileName;
+  }
+
+  // From https://developer.android.com/privacy-and-security/risks/path-traversal#path-traversal-mitigations.
+  protected static @NonNull File saferOpenFile(@NonNull String path, @NonNull String expectedDir) throws IllegalArgumentException, IOException {
+    File f = new File(path);
+    String canonicalPath = f.getCanonicalPath();
+    if (!canonicalPath.startsWith(expectedDir)) {
```

Comment:
In theory, this check would allow a path traversal into a directory that has `expectedDir` as a prefix.
```java
String expectedDir = "/data/data/com.app/cache/123";
String path = "/data/data/com.app/cache/123/../12345/traversal.png";
String canonicalPath = "/data/data/com.app/cache/12345/traversal.png";
canonicalPath.startsWith(expectedDir); // true
```
However, because of the filename sanitization earlier, path traversal sequences are already prevented and this cannot occur. Additionally, the expected directory is an unknown UUID. So, this is not an issue here; I just wanted to mention it for next time. :)

Reply:
I think this would be an interesting comment to convey to the authors of that method/maintainers of the [page](https://developer.android.com/privacy-and-security/risks/path-traversal), I can try to bring it to their attention. Presumably this would be fixed by making the check instead be
```java
canonicalPath.startsWith(expectedDir + "/")
```
(after optionally stripping a `/` char at the very end so that we don't end up expecting `"some/path/"` + `"/"`)

Reply:
I found a specific GitHub CodeQL docs page about the Partial Path Traversal issue that explains a better mitigation using `Path` objects
https://codeql.github.com/codeql-query-help/java/java-partial-path-traversal-from-remote/